File: src/services/stock-edit-request.service.ts
import api from "./api";
import { ApiResponse, PaginatedResponse } from "@/types/api.types";
import {
  CreateStockEditRequestPayload,
  ReviewStockEditRequestPayload,
  StockEditRequest,
  StockEditRequestFilters,
  StockEditRequestSummary,
} from "@/types/stock-edit-request.types";

function buildParams(filters?: StockEditRequestFilters) {
  if (!filters) return undefined;
  const out: Record<string, string | number> = {};
  if (filters.page !== undefined) out.page = filters.page;
  if (filters.size !== undefined) out.size = filters.size;
  if (filters.search) out.search = filters.search;
  if (filters.status) out.status = filters.status;
  if (filters.productId !== "" && filters.productId !== undefined) out.productId = filters.productId;
  if (filters.warehouseId !== "" && filters.warehouseId !== undefined) out.warehouseId = filters.warehouseId;
  if (filters.dateFrom) out.dateFrom = filters.dateFrom;
  if (filters.dateTo) out.dateTo = filters.dateTo;
  return out;
}

export const stockEditRequestService = {
  list: (filters?: StockEditRequestFilters) =>
    api.get<PaginatedResponse<StockEditRequest>>("/stock-edit-requests", {
      params: buildParams(filters),
    }),

  getById: (id: number) =>
    api.get<ApiResponse<StockEditRequest>>(`/stock-edit-requests/${id}`),

  summary: () =>
    api.get<ApiResponse<StockEditRequestSummary>>("/stock-edit-requests/summary"),

  create: (data: CreateStockEditRequestPayload) =>
    api.post<ApiResponse<StockEditRequest>>("/stock-edit-requests", data),

  approve: (id: number, data: ReviewStockEditRequestPayload) =>
    api.post<ApiResponse<StockEditRequest>>(`/stock-edit-requests/${id}/approve`, data),

  reject: (id: number, data: ReviewStockEditRequestPayload) =>
    api.post<ApiResponse<StockEditRequest>>(`/stock-edit-requests/${id}/reject`, data),

  cancel: (id: number) =>
    api.post<ApiResponse<StockEditRequest>>(`/stock-edit-requests/${id}/cancel`, {}),

  comment: (id: number, data: ReviewStockEditRequestPayload) =>
    api.post<ApiResponse<StockEditRequest>>(`/stock-edit-requests/${id}/comment`, data),
};

File: inflow-backend/src/main/java/com/john/inflow/service/StockEditRequestService.java
package com.john.inflow.service;

import com.john.inflow.dto.request.CreateStockEditRequest;
import com.john.inflow.dto.request.ReviewStockEditRequest;
import com.john.inflow.dto.response.PageResponse;
import com.john.inflow.dto.response.StockEditRequestResponse;
import com.john.inflow.dto.response.StockEditRequestSummaryResponse;
import com.john.inflow.entity.User;

import java.time.OffsetDateTime;

public interface StockEditRequestService {
    StockEditRequestResponse create(CreateStockEditRequest request, User actor);
    StockEditRequestResponse getById(Integer id, User actor);
    PageResponse<StockEditRequestResponse> search(
            String search,
            String status,
            Integer productId,
            Integer warehouseId,
            OffsetDateTime dateFrom,
            OffsetDateTime dateTo,
            int page,
            int size,
            User actor
    );
    StockEditRequestResponse approve(Integer id, ReviewStockEditRequest request, Integer userId);
    StockEditRequestResponse reject(Integer id, ReviewStockEditRequest request, Integer userId);
    StockEditRequestResponse cancel(Integer id, Integer userId);
    StockEditRequestResponse addComment(Integer id, ReviewStockEditRequest request, Integer userId);
    StockEditRequestSummaryResponse getSummary(User actor);
}
